src/code/middleware/state-save.js

```diff
@@ -0,0 +1,29 @@
+import actionTypes from '../action-types';
+import { getClassId, getUserId } from '../utilities/url-params';
+
+export const authoringVersionNumber = 1;
+
+export default () => store => next => action => {
+  let result = next(action);
+
+  if (action.type === actionTypes.CHALLENGE_COMPLETED) {
+      let userQueryString = getUserQueryString();
+
+      if (userQueryString) {
+        let nextState = store.getState(),
+            gems = nextState.gems,
+            stateUpdate = {state: gems};
+          
+        firebase.database().ref(userQueryString).update(stateUpdate); //eslint-disable-line
+      }
+    }
+
+  return result;
+};
+
+export function getUserQueryString() {
```

Comment:
I wasn't sure where to put utilities like this - they're in the middleware for now, but we may want them somewhere else if we build out our database interactions